lib/licence_location_identifier.rb

```diff
@@ -5,27 +5,44 @@ class LicenceLocationIdentifier < LocationIdentifier
   # to find the most appropriate council from the "providing tier"
   # if not, then return the closest authority available.
   def self.find_slug(areas, artefact = nil)
-    councils = areas.map do |area|
-      [area["type"], area["codes"] && area["codes"]["govuk_slug"]]
-    end
+    new(areas, artefact).find_slug
+  end
 
-    if artefact and providing_tier = artefact["details"].try(:[], "licence").try(:[], "local_service").try(:[], "providing_tier")
-      councils_by_tier = Hash[councils.collect do |k, v|
-        [self.identify_tier(k), v]
-      end]
+  attr_reader :areas, :artefact
 
-      appropriate_authority = providing_tier.map do |tier|
-        councils_by_tier[tier]
-      end.compact.first
+  def initialize(areas, artefact)
+    @areas = areas
+    @artefact = artefact
+  end
 
-      return appropriate_authority unless appropriate_authority.nil?
+  def find_slug
+    if providing_tier && matching_authority_by_tier
+      matching_authority_by_tier.try(:[], "codes").try(:[], "govuk_slug")
+    elsif matching_authority_by_area_type
+      matching_authority_by_area_type.try(:[], "codes").try(:[], "govuk_slug")
     end
```

Comment:
Another difference between the old and new implementations is that in the `if providing_tier` conditional in the old version our result could be `nil` if we don't have an area that maps to the providing tier *or* if we do have one but it has no slug.  If the result of that conditional is `nil` we fall out of the conditional and try the other way of getting a slug.  In the new implementation we'd only fall through if we couldn't find an area in `matching_authority_by_tier`, *not* if we do find one but it has no slug.

I'm less sure this is a problem based on the data we actually use, but still, edgecases are what caused bugs.